ISPN-6405 Persistence configuration with clustered cache can cause

wburns opened the pull request:
duplicate expiration messages

* Fixed issue where reaper caused 2 expirations in cluster cache

https://issues.jboss.org/browse/ISPN-6405

slaskawi commented:
Looks good to me

slaskawi commented:
Could you please rebase it? I would like to make sure those failures will go away...

wburns commented:
Rebased

danberindei commented on core/src/main/java/org/infinispan/commands/write/RemoveExpiredCommand.java:
```diff
@@ -118,8 +118,8 @@ public byte getCommandId() {
-              "value=" + value +
-              "lifespan=" + lifespan +
+              " value=" + value +
+              " lifespan=" + lifespan +
```
Might as well add a comma

wburns replied:
Sure

danberindei commented on core/src/test/java/org/infinispan/expiration/impl/ExpirationSingleFileStoreDistListenerFunctionalTest.java:
```diff
@@ -0,0 +1,26 @@
+      config
+              // We set the wakeup interval to max value.  If we have a store we force the reaper to be enabled, but
+              // for this test we can get false positives if the reaper runs at the same time as a manual expiration
+              // process
```
How about "Prevent the reaper from running, reaperEnabled(false) doesn't work when a store is present"?

wburns replied:
okay

wburns commented:
Updated with the comments @danberindei.  I didn't squash the commit so you can see the diff easily.  Still needs to be fixed up on a rebase.

danberindei commented on core/src/test/java/org/infinispan/expiration/impl/ExpirationStoreListenerFunctionalTest.java:
```diff
@@ -56,18 +57,20 @@ public void testExpirationOfStoreWhenDataNotInMemory() throws Exception {
 
    private void assertExpiredEvents(int count) {
-      assertEquals(count, listener.getInvocationCount());
+      eventually("Last count was :" + listener.getInvocationCount(), () -> count == listener.getInvocationCount());
```
We have `eventuallyEquals` for this

wburns replied:
I didn't realize the trace message was what I wanted. Thanks.

wburns commented:
Comments address, still will need to be rebased before integration.

wburns commented:
Something changed that is causing this to fail now, not sure what since nothing has changed in this PR by itself that would have caused it, looking..

wburns commented:
Looks like the failures is caused by https://github.com/infinispan/infinispan/commit/d60bdc6ad4cb62772dade4b12afd92cddd0a6af0#diff-719e44d55c697435cc2ae65da78d4a35 since the purge is no longer synchronous.

wburns commented:
I don't know how I didn't see the failures in my tests before, but it is caused due to the SingleFileStore purging entries asynchronously.  And to be honest there is no way to really fix this in a nice way, besides tests.

The API doesn't define a Future or anything for the caller to return to block to wait on.  This would mean that could theoretically get a pile up of expiration threads if there was a low reaper wait time and an expiration for a cache store that would take a while.

wburns commented:
Will just make the calling thread block until the expiration thread is completed.  Unfortunately this means we are using a thread for nothing, while it could have just been processing the expired entries directly.

danberindei commented:
@wburns I see some of the stores use a `ExecutorCompletionService` to make sure the purge is done by the time the method returns, we could do the same in `SingleFileStore`. And we should make it clear in the `purge()` javadoc that we expect it to be finished when it returns.

We could enforce this by wrapping the persistence executor and counting started/finished tasks in the wrapper, but I'm not sure if we should do it in production or just for these tests.

wburns commented:
@danberindei That is what I ended up doing, re: the `ExecutorAllCompletionService`.  I will add the javadoc mention though.

wburns commented:
Added another commit that should fix that all up, let me know what you think @danberindei

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -663,47 +666,53 @@ private void mergeFreeEntries(List<FileEntry> entries) {
+                  } catch (Exception e) {
+                     throw new PersistenceException(e);
                   }
+                  if (task != null) task.entryPurged(next.getKey());
```
I'd rather run only the PurgeListener on the Executor.

wburns replied:
I had thought about that.  My worry is that the listener may not do anything, so it would end up causing a context switch with no benefit.  I am thinking I can submit them in batches using a parallel stream though.  Check it out after I resubmit to let me know what you think.

wburns commented:
Updated to use persistence thread pool for the notifications and rebased.

danberindei commented on core/src/main/java/org/infinispan/executors/ParallelExecutorHelper.java:
```diff
@@ -0,0 +1,59 @@
+      long size = spliterator.getExactSizeIfKnown();
+      // If the size is unknown we use the builder method to prevent multiple resizings
+      if (size == -1) {
+         // Use stream builder as it requires less resizing than ArrayList if the spliterator is large
```
I don't think keeping all the entries/tasks in memory is a good idea anyway, even if it requires less resizing.

danberindei replied:
I don't think keeping all the entries/tasks in memory is a good idea anyway, even if it requires less resizing.
That's the reasoning behind `ExecutorAllCompletionService`, to avoid keeping all the futures in memory.

wburns replied:
Hrmm, when the `ExecutorAllCompletionService` was added it was more about not having to keep track of # of executed tasks in the user code.  The futures not being retained for the duration was just a side effect.

And the comment here is not talking about ExecutorAllCompletionService but rather the fact it is using a Stream.Builder instead of an ArrayList.

I can try to convert this over to use the EACS, but I am not sure if there is a nice way to also return a `CompletableFuture` as well since I won't have a thread to complete the Future when all tasks are completed since there is no guarantee that a user thread is available to check it.

danberindei commented on core/src/main/java/org/infinispan/executors/ParallelExecutorHelper.java:
```diff
@@ -0,0 +1,59 @@
+   private static <T> CompletableFuture<Void> forEach(Spliterator<? extends T> spliterator,
+           Consumer<? super T> consumer, Executor executor, Consumer<CompletableFuture<Void>> builder) {
+      Spliterator<? extends T> split;
+      while ((split = spliterator.trySplit()) != null) {
```
Shouldn't you stop splitting when the spliterator reaches a certain size to batch consumer calls? Looking at the `ArrayListSpliterator` code, it only stops splitting when it has a single element.

wburns replied:
Hrmm I must be misremembering how the spliterators are implemented. I will
take a closer look and remedy.

On Tue, Apr 5, 2016, 4:29 AM Dan Berindei <notifications@github.com> wrote:

> In core/src/main/java/org/infinispan/executors/ParallelExecutorHelper.java
> <https://github.com/infinispan/infinispan/pull/4160#discussion_r58503539>:
>
> > +      } else {
> > +         if (size > (long) Integer.MAX_VALUE) {
> > +            throw new IllegalArgumentException("Spliterator has " + size +
> > +                    " entries, we can't support more than Integer.MAX_VALUE");
> > +         }
> > +         // Use stream builder as it requires less resizing than ArrayList if the spliterator is large
> > +         List<CompletableFuture<Void>> list = new ArrayList<>((int) size);
> > +         list.add(forEach(spliterator, consumer, executor, list::add));
> > +         return CompletableFuture.allOf((CompletableFuture<Void>[]) list.toArray(new CompletableFuture[list.size()]));
> > +      }
> > +   }
> > +
> > +   private static <T> CompletableFuture<Void> forEach(Spliterator<? extends T> spliterator,
> > +           Consumer<? super T> consumer, Executor executor, Consumer<CompletableFuture<Void>> builder) {
> > +      Spliterator<? extends T> split;
> > +      while ((split = spliterator.trySplit()) != null) {
>
> Shouldn't you stop splitting when the spliterator reaches a certain size
> to batch consumer calls? Looking at the ArrayListSpliterator code, it
> only stops splitting when it has a single element.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/infinispan/infinispan/pull/4160/files/77664363ebd5e99b6a689093b63cff4ae12eb740#r58503539>
>

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -35,33 +37,33 @@
+ * cache where loss of data in the cache store does not lead to data loss, and data can be recomputed or re-queried from
+ * the original data source.
+ * <p>
+ * This class is fully thread safe, yet allows for concurrent load / store of individual cache entries.
```
Did you actually change anything here?

danberindei commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -219,8 +220,8 @@ private void rebuildIndex() throws Exception {
-    * The base class implementation calls {@link #load(Object)} for this, we can do better because
-    * we keep all keys in memory.
+    * The base class implementation calls {@link #load(Object)} for this, we can do better because we keep all keys in
+    * memory.
```
Better configure IntelliJ to only reformat changed lines :)

wburns replied:
Yeah I noticed this too late. I can fix this and put it in another commit.

On Tue, Apr 5, 2016, 4:30 AM Dan Berindei <notifications@github.com> wrote:

> In core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java
> <https://github.com/infinispan/infinispan/pull/4160#discussion_r58503625>:
>
> > @@ -219,8 +220,8 @@ private void rebuildIndex() throws Exception {
> >     }
> >
> >     /**
> > -    * The base class implementation calls {@link #load(Object)} for this, we can do better because
> > -    * we keep all keys in memory.
> > +    * The base class implementation calls {@link #load(Object)} for this, we can do better because we keep all keys in
> > +    * memory.
>
> Better configure IntelliJ to only reformat changed lines :)
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/infinispan/infinispan/pull/4160/files/77664363ebd5e99b6a689093b63cff4ae12eb740#r58503625>
>

wburns commented:
Okay I have just changed it to not use the thread pool with the single file store for now.  Also I added in ISPN-6071 since it needs the new test that I have with a slight change.

wburns commented:
We can cherry pick this into 8.2.x, but I will make a separate PR after this is integrated for 8.1.x for ISPN-6071.

danberindei commented:
@wburns Some new test failures:
http://ci.infinispan.org/viewLog.html?buildId=38345&buildTypeId=bt9&guest=1

wburns commented:
@danberindei I have rebased and I haven't been able to reproduce this issue at all yet.

tristantarrant commented:
Pushed, thanks. Can we have an 8.2.x fix please ?
